File: AtCoder/ABC/ABC051-100/abc073/abc073d.cpp
#include <bits/stdc++.h>
using namespace std;
const long long INF = 1LL<<60;
#define REP(i,b,e) for(int i=(b);i<(e);++i)
#define rep(i,e) for(int i=0;i<(e);++i)

long long dist[200][200];
int N, M, R, a, b, c, rt;

int main() {
  cin >> N >> M >> R;
  vector<int> r;

  rep(i, 200)rep(j, 200){
    if(i==j) dist[i][j] = 0;
    else dist[i][j] = INF;
  }

  rep(i, R){
    cin >> rt;
    r.push_back(--rt);
  }
  rep(i, M){
    cin >> a >> b >> c;
    a--; b--;
    dist[a][b] = c;
    dist[b][a] = c;
  }

  rep(k, N)rep(i, N)rep(j, N){
    dist[i][j] = min(dist[i][j], dist[i][k]+dist[k][j]);
  }

  long long ans = INF;
  sort(r.begin(), r.end());
  do {
    long long cur = 0;
    rep(i, R-1) cur += dist[r[i]][r[i+1]];
    ans = min(ans, cur);
  } while (next_permutation(r.begin(), r.end()));

  cout << ans << endl;
  return 0;
}

File: AtCoder/ABC/ABC051-100/abc073/abc073d.py
n, m, R = list(map(int, input().split()))
dist = [[float("INF") if i!=j else 0 for i in range(n)] for j in range(n)]
r = list(map(int, input().split()))

for i in range(m):
  a, b, c = list(map(int, input().split()))
  a, b = a-1, b-1
  dist[a][b] = c
  dist[b][a] = c

for k in range(n):
  for i in range(n):
    for j in range(n):
      dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

from itertools import permutations
ans = float("INF")
for e in permutations(r):
  cur = 0
  for i in range(R-1):
    cur += dist[e[i]-1][e[i+1]-1]
  ans = min(ans, cur)
print(ans)